llvm/lib/Transforms/Vectorize/VectorCombine.cpp

```diff
@@ -1547,6 +1548,78 @@ bool VectorCombine::foldShuffleOfCastops(Instruction &I) {
   return true;
 }
 
+/// Try to convert "shuffle (shuffle x, undef), (shuffle y, undef)"
+/// into "shuffle x, y".
+bool VectorCombine::foldShuffleOfShuffles(Instruction &I) {
+  Value *V0, *V1;
+  ArrayRef<int> OuterMask, InnerMask0, InnerMask1;
+  if (!match(&I, m_Shuffle(m_OneUse(m_Shuffle(m_Value(V0), m_Undef(),
+                                              m_Mask(InnerMask0))),
+                           m_OneUse(m_Shuffle(m_Value(V1), m_Undef(),
+                                              m_Mask(InnerMask1))),
+                           m_Mask(OuterMask))))
+    return false;
+
+  auto *ShuffleDstTy = dyn_cast<FixedVectorType>(I.getType());
+  auto *ShuffleSrcTy = dyn_cast<FixedVectorType>(V0->getType());
+  auto *ShuffleImmTy = dyn_cast<FixedVectorType>(I.getOperand(0)->getType());
+  if (!ShuffleDstTy || !ShuffleSrcTy || !ShuffleImmTy ||
+      V0->getType() != V1->getType())
+    return false;
+
+  unsigned NumSrcElts = ShuffleSrcTy->getNumElements();
+  unsigned NumImmElts = ShuffleImmTy->getNumElements();
+
+  // Bail if either inner masks reference the RHS undef/poison arg.
+  if (any_of(InnerMask0, [&](int M) { return M >= (int)NumSrcElts; }) ||
+      any_of(InnerMask1, [&](int M) { return M >= (int)NumSrcElts; }))
+    return false;
```

Comment:
Referencing poison is not a problem (can be just ignored), but it might be a problem for undef

Reply:
The problem isn't whether its undef/poison - its that we'd be creating a bad mask index below in NewMask (especially for InnerMask1 which would be out of bounds).

Reply:
But if you select the index for poison, you can easily treat it as PoisonMaskElem, just replace it with PoisonMaskElem in the mask.